WIP: ITK to DREAM3D filters

fbudin69500 opened the pull request.

fbudin69500 committed:
ENH: Conversion of ITK Images to Dream3D DataContainer

Addition of filters to convert ITK images to Dream3D DataContainers
and Dream3D DataContainers to ITK Images. Tests have been implemented
to test the current API.

fbudin69500 committed:
STYLE: Remove extra spaces

Visual Studio adds extra spaces, when starting a namespace,
that needs to be removed.

fbudin69500 commented:
@thewtex @vovythevov Here is the PR.

thewtex commented on ITKImageProcessingFilters/itkDream3DDataToImageFilter.h:
```diff
@@ -0,0 +1,45 @@
+		typedef SmartPointer<Self>                                         Pointer;
+		typedef typename Image<PixelType, VDimension> ImageType;
+		typedef typename ImageType::Pointer ImagePointer;
+		typedef typename DataArray<PixelType> DataArrayPixelType;
```
please align typedefs

thewtex commented on ITKImageProcessingFilters/itkDream3DDataToImageFilter.h:
```diff
@@ -0,0 +1,45 @@
+		typedef typename DataArray<PixelType> DataArrayPixelType;
+		/** Method for creation through the object factory. */
+		itkNewMacro(Self);
+		void SetInput(DataContainer::Pointer dc);
```
virtual?

thewtex commented on ITKImageProcessingFilters/itkDream3DDataToImageFilter.h:
```diff
@@ -0,0 +1,45 @@
+		Dream3DDataToImageFilter();
+		virtual ~Dream3DDataToImageFilter();
+		void Check();
+        void GenerateOutputInformation();
```
virtual

thewtex replied:
also use ITK_OVERRIDE at the end.

thewtex commented on ITKImageProcessingFilters/itkDream3DDataToImageFilter.h:
```diff
@@ -0,0 +1,45 @@
+		virtual ~Dream3DDataToImageFilter();
+		void Check();
+        void GenerateOutputInformation();
+        void GenerateData();
```
virtual + ITK_OVERRIDE

thewtex commented on ITKImageProcessingFilters/itkDream3DDataToImageFilter.hxx:
```diff
@@ -0,0 +1,116 @@
+template< typename PixelType, unsigned int VDimension>
+void
+Dream3DDataToImageFilter< PixelType, VDimension >
+::SetInput(DataContainer::Pointer dc)
```
Please pass a reference:

```
::SetInput(DataContainer::Pointer & dc)
```

thewtex commented on ITKImageProcessingFilters/itkDream3DDataToImageFilter.hxx:
```diff
@@ -0,0 +1,116 @@
+template< typename PixelType, unsigned int VDimension>
+void
+Dream3DDataToImageFilter< PixelType, VDimension >
+::Check()
```
cool :+1:

thewtex commented on ITKImageProcessingFilters/itkDream3DDataToImageFilter.hxx:
```diff
@@ -0,0 +1,116 @@
+	this->SetSpacing(spacing);
+	this->SetRegion(size);
+	this->SetDirection(direction);
+	ImportImageFilter::GenerateOutputInformation();
```
It is minor, but it is a best practice to use Superclass::GenerateOutputInformation(), which prevents bugs if the base class is changed.

thewtex commented on ITKImageProcessingFilters/itkDream3DDataToImageFilter.hxx:
```diff
@@ -0,0 +1,116 @@
+
+template< typename PixelType, unsigned int VDimension>
+void
+Dream3DDataToImageFilter< PixelType, VDimension >::GenerateData()
```
Please put `::GenerateData()` on a separate line

thewtex commented on ITKImageProcessingFilters/itkDream3DDataToImageFilter.hxx:
```diff
@@ -0,0 +1,116 @@
+	this->SetImportPointer(static_cast<float*>(dataArray->getVoidPointer(0)), dataArray->getSize(),
+		importImageFilterWillOwnTheBuffer);
+	ImportImageFilter::GenerateData();
+}
```
Nice! The inheritance from ImportImageFilter is a good approach.

thewtex commented on ITKImageProcessingFilters/itkImageToDream3DDataFilter.h:
```diff
@@ -0,0 +1,46 @@
+		void Update();
+		DataContainer::Pointer GetOutput();
+	protected:
+		void Check();
```
For this and the other class, a more descriptive name for `Check` could help.  Maybe `ValidInputCheck` or ...

thewtex commented on ITKImageProcessingFilters/itkImageToDream3DDataFilter.hxx:
```diff
@@ -0,0 +1,141 @@
+{
+	if (this->GetInput(0) && this->GetMTime() > m_OutputUpdated )
+	{
+		this->Update();
```
This could be removed?

thewtex commented:
Digging in a bit more, it would be good if the `itk::DREAM3DDataToImageFilter` generated an `itk::DREAM3DImportedImage` output, where `itk::DREAM3DImportedImage` inherits from `itk::Image` but also has a std::shared_ptr to the DREAM3D pixel data object.

For `itk::DREAM3DDataToImageFilter`, it may be helpful to also look at at `itk::ImageSource` code as a reference.  Like `itk::ImageSource`, `GetOutput()` and `MakeOutput` could be implemented.

fbudin69500 commented:
Thanks for the input!

fbudin69500 committed:
ENH: Following more closely ITK's API

In itkImageToDream#DDataFilter, the data container is embedded
in itkSimpleDataObjectDecorator which allows to really use ITK's API.
A few style improvements have also been done (typedef alignment,
passing arguments by reference, better function naming). Renaming
ValidInputCheck() to VerifyPreconditions() which is a virtual function
in itkProcessObject. This allows to not have to manage the validation
of the inputs manually in itkImageToDream3DDataFilter. Additionally,
GenerateData() was split in two functions: GenerateData() and
GenerateOutputInformation() to replicate ITK's API.

fbudin69500 committed:
ENH: itkImageToDream3DDataFilter takes a data container as input

itkImageToDream3DDataFilter does not create the data container
anymore. It accepts a data container as an input and adds the
image geometry (if necessary), the attribute matrix (if necessary),
and the data array to the given data container.

fbudin69500 commented:
@thewtex I have added the ITK decorator to itkImageToDream3DDataFilter. I have also modified the behavior of the filter so that it does not create a DREAM3D data container anymore, but requires it as an input. I have not worked on the memory management yet.

thewtex commented:
@fbudin69500 awesome!

thewtex commented on ITKImageProcessingFilters/itkImageToDream3DDataFilter.hxx:
```diff
@@ -136,6 +147,7 @@ ImageToDream3DDataFilter<PixelType, VDimension>
 	{
 		itkExceptionMacro("DataArrayName contains a '/'");
 	}
+	Superclass::VerifyPreconditions();
```
Nice

fbudin69500 committed:
BUG: PixelType was hardcoded to float

fbudin69500 committed:
ENH: New tests verifying that data is transfered after filter

Addition of tests that make sure that when converting Dream3D data
to ITK data and ITK data to Dream3D data, the resulting data
container belongs to the new object, and is not destroyed when the
original object is destroyed.

fbudin69500 committed:
BUG: Missing 'typename' declarations

Code was compiling with Visual Studio C++ 2013 but not with Clang.

thewtex commented:
Here is the build error:

  http://my.cdash.org/viewBuildError.php?buildid=964172

Maybe SIMPLView needs to be updated in the Docker image or on your local system?

fbudin69500 committed:
Updating code due to new organization of SIMPLView repository [1]

[1] https://github.com/BlueQuartzSoftware/DREAM3D/commit/2a279c23238304b78aa97b91ca1a2d39e27b204a

fbudin69500 commented:
@thewtex Actually, SIMPLView had been updated in DREAM3D and it required updating a few lines of code of ITKImageProcessing.

fbudin69500 committed:
ENH: Addition of in-place filters

Creation of in-place filters to convert Dream3D data to ITK data, and ITK
data to Dream3D data. These new filters are derived from the original
filters that will copy the data from the original container to a new
memory location.

fbudin69500 committed:
ENH: Creation of data conversion filters that allocate memory for output.

The new Dream3D data to ITK image and ITK image to Dream3D data conversion
filters allocate memory for their output. They do not compute the result in
place.

fbudin69500 committed:
BUG: InPlace  filters release ownership from input

In inplace filters, ownership of the data is transfrered from the input to
the output data container. Input data containers still point to the data,
but will not release the data when they are deleted.

fbudin69500 committed:
STYLE: Remove tabulations and improve code alignment

fbudin69500 commented:
@thewtex @vovythevov Please review. I have added both the In-place and "not in-place" (memory allocation) filters. You will need the path that I am about to commit for ITK for the tests to pass.

fbudin69500 commented:
ITK patch submitted: http://review.source.kitware.com/#/c/21204/

fbudin69500 committed:
ENH: Merged "inPlace" filters with "not inPlace" filters

To simplify the code, and to follow the example from
itkInPlaceImageFilter, the different classes ("inPlace" and "not inPlace")
were merged into the "inPlace" filters. An option has been added to select
if the processing is done in place or not.

thewtex commented on ITKImageProcessingFilters/itkInPlaceDream3DDataToImageFilter.h:
```diff
@@ -1,32 +1,46 @@
+    itkGetMacro(DataArrayName, std::string);
+    itkSetMacro(MatrixArrayName, std::string);
+    itkGetMacro(MatrixArrayName, std::string);
+    itkSetMacro( InPlace, bool );
```
Cool. Nice to keep the same of the member as the InPlaceImageFilter.

thewtex commented on ITKImageProcessingFilters/itkInPlaceDream3DDataToImageFilter.hxx:
```diff
@@ -19,16 +22,107 @@ InPlaceDream3DDataToImageFilter<PixelType, VDimension>::~InPlaceDream3DDataToIma
+  this->SetSpacing(spacing);
+  this->SetRegion(size);
+  this->SetDirection(direction);
+  Superclass::GenerateOutputInformation();
```
This could go at the beginning of the method so that it does not overwrite any changes above.

fbudin69500 replied:
Actually Superclass::GenerateOutputInformation() needs the origin, spacing, size, and direction information to be set before it is called, as it uses these values to set the output image properties. So it must be called last.

thewtex replied:
Yes, looking at itkImportImageFilter.hxx, you are right.

But, examining the ImportImageFilter class suggests that we should not inherit from it, but itk::ImageSource instead. The SetSpacing, SetOrigin, SetDirections are no longer valid since this taken from the DREAM.3D ImageGeom.

thewtex commented on ITKImageProcessingFilters/itkInPlaceImageToDream3DDataFilter.hxx:
```diff
@@ -10,6 +12,36 @@ template<typename PixelType, unsigned int VDimension>
+  // Create the output. We use static_cast<> here because we know the default
+  // output must be of type DecoratorType
+  typename DecoratorType::Pointer output =
+    static_cast< DecoratorType * >(this->MakeOutput(0).GetPointer());
```
Nice

thewtex commented:
Nice! This is very well done.

The ITK patch has been merged.  I will update the Docker image...

fbudin69500 commented:
@thewtex Thanks for the review. I commented your comment :)

fbudin69500 committed:
ENH: itkInPlaceDream3DDataToImageFilter derives from ImageSource

itkInPlaceDream3DDataToImageFilter was deriving from itkImportImageFilter.
However, itkImportImageFilter contains defines function to set the
spacing, the origin, and the direction of the output image. These
functions should not be instantiated for
itkInPlaceDream3DDataToImageFilter since these image properties are set by
the Dream3D data container.

fbudin69500 commented:
@thewtex Code updated to take into account your comments. Thanks!
@vovythevov Feel free to review the code, or use it as it should be stable now.

thewtex commented:
@fbudin69500 Nice! Looks good!

The CircleCI configuration may need to have the testing step timeout increased a la

  https://github.com/KitwareMedical/TubeTK/blob/e19f725ecac22a9a290d9c093fbff71b4ef294bf/circle.yml#L15

fbudin69500 commented:
Maybe, but the tests are fast on my machine (a few second each, and only about 20 tests) so maybe the problem is somewhere else. It seems that there are some compilation issues that need to be resolved, maybe by updating the docker image?

thewtex commented:
Since CircleCI may have a different version of the base Docker image, it may try to build it itself.

fbudin69500 committed:
ENH: Increase CircleCI timeout

Increase CircleCI timeout to 6000s.

fbudin69500 commented:
@thewtex Good catch for the timeout. Now it works :+1:

thewtex commented:
@fbudin69500 way to go! :+1:

imikejackson commented on ITKImageProcessingFilters/itkInPlaceDream3DDataToImageFilter.h:
```diff
@@ -0,0 +1,57 @@
+    virtual void SetInput(DataContainer::Pointer &dc);
+    itkSetMacro(DataArrayName, std::string);
+    itkGetMacro(DataArrayName, std::string);
+    itkSetMacro(MatrixArrayName, std::string);
```
Can you change MatrixArrayName to AttributeMatrixArrayName to be consistent with DREAM.3D naming conventions

imikejackson commented on ITKImageProcessingFilters/itkInPlaceDream3DDataToImageFilter.h:
```diff
@@ -0,0 +1,57 @@
+    itkTypeMacro(InPlaceDream3DDataToImageFilter, ImportImageFilter);
+    virtual void SetInput(DataContainer::Pointer &dc);
+    itkSetMacro(DataArrayName, std::string);
+    itkGetMacro(DataArrayName, std::string);
```
Can you also add API to take a DataArrayPath object as a convenience method to set the AttributeMatrixArrayName and the DataArrayName in one method.

fbudin69500 replied:
Addressed.

imikejackson commented on ITKImageProcessingFilters/itkInPlaceImageToDream3DDataFilter.hxx:
```diff
@@ -0,0 +1,240 @@
+  if (ma->doesAttributeArrayExist(m_DataArrayName.c_str()))
+  {
+    itkExceptionMacro("Existing data array with same name.");
+  }
```
Could we add a boolean member variable to this class that determines whether this check is run?  During a DREAM.3D filter's preflight, we will often stick arrays into the data structure to show up in the user interface so that the user knows what arrays will be created/removed; in this special case, we'll want to create 'dummy' arrays during this step that get overwritten later on by the itkInPlaceImageToDream3DDataFilter class.  Having an option here to either do this check or not gives us the flexibility to operate either way.

fbudin69500 replied:
Maybe we can wait until this case arises? I would think that if the code reaches this point and tests are not satisfied, I would expect the function to crash. Maybe in preflight, the filter should not be updated, and therefore this piece of code would not be execute. What do you think?

imikejackson replied:
]

Francois Budin wrote:

> In ITKImageProcessingFilters/itkInPlaceImageToDream3DDataFilter.hxx
> https://github.com/BlueQuartzSoftware/ITKImageProcessing/pull/4#discussion_r66332181:
> 
> > -      itkExceptionMacro("Tuples dimension of existing matrix array do not match image size.");
> > -    }
> > -    if (ma->getType() != SIMPL::AttributeMatrixType::Cell)
> > -    {
> > -      itkExceptionMacro("Attribute matrix is not of type SIMPL::AttributeMatrixType::Cell.");
> > -    }
> > -  }
> > -  else
> > -  {
> > -    ma = dataContainer->createAndAddAttributeMatrix(tDims, m_MatrixArrayName.c_str(), SIMPL::AttributeMatrixType::Cell);
> > -  }
> > -  // Checks if doesAttributeArray exists
> > -  if (ma->doesAttributeArrayExist(m_DataArrayName.c_str()))
> > -  {
> > -    itkExceptionMacro("Existing data array with same name.");
> > -  }
> 
> Maybe we can wait until this case arises? I would think that if the code
> reaches this point and tests are not satisfied, I would expect the
> function to crash. Maybe in preflight, the filter should not be updated,
> and therefore this piece of code would not be execute. What do you think?

During the "dataCheck()" function that is executed during the 
preflight() stage any created arrays will be added to the 
DataContainer*. This is done so that filters next in the pipeline are 
aware of the complete structure at that point in the pipeline. During 
the preflight process the DataArrays only have their meta-data 
initialized but the actual memory that holds the data is not allocated 
but the AttributeMatrix will still have an AttributeArray with the 
designated ArrayName. This means that when running the filter from any 
of our GUI applications that bit of code will evaluate to true which 
means we will hit the exception in your code.

  There are a few exceptions to this rule where during the "execute" 
function we remove the DataArray from the AttributeMatrix then insert a 
new one that has the same name. This is the exception and not the rule. 
I think if we can be more consistent throughout the DREAM.3D code base 
it would be better.

  Let's get your fork merged into the main repository and I can help 
write an actual filter that does something basic using ITK (like a 
Gaussian blur). I think this will help explain our desires for the API.

imikejackson commented:
Also add near the bottom of the ItkImageProcessingFilter/SourceList.cmake the following:

#-------------

# These are files that need to be compiled into DREAM3DLib but are NOT filters

ADD_SIMPL_SUPPORT_HEADER(${${PLUGIN_NAME}_SOURCE_DIR} ${_filterGroupName} itkInPlaceImageToDream3DDataFilter.h)
ADD_SIMPL_SUPPORT_HEADER(${${PLUGIN_NAME}_SOURCE_DIR} ${_filterGroupName} itkInPlaceImageToDream3DDataFilter.hxx)
ADD_SIMPL_SUPPORT_HEADER(${${PLUGIN_NAME}_SOURCE_DIR} ${_filterGroupName} itkInPlaceDream3DDataToImageFilter.h)
ADD_SIMPL_SUPPORT_HEADER(${${PLUGIN_NAME}_SOURCE_DIR} ${_filterGroupName} itkInPlaceDream3DDataToImageFilter.hxx)

just above the following CMake code:
#---------------------

# This macro must come last after we are done adding all the filters and support files.

SIMPL_END_FILTER_GROUP(${${PLUGIN_NAME}_BINARY_DIR} "${_filterGroupName}" "${PLUGIN_NAME}")

This will add those headers to IDE's like Visual Studio, Xcode and QtCreator.

fbudin69500 committed:
ENH: Adds ITK to DREAM3D conversion filters to project for IDE

fbudin69500 committed:
ENH: Follow DREAM3D naming conventions

Renamed MatrixArrayName to AttributeMatrixArrayName and
added a function to set AttributeMatrixArrayName and DataArrayName
from DataArrayPath.

fbudin69500 commented:
@imikejackson I pushed the modifications.

imikejackson commented:
Guys,
    I was implementing a test filter as we talked about on the hangout 
earlier today and I came across this bit of code from another 
implementation:

void itk::ImportImageContainer< TElementIdentifier, TElement 
 >::SetImportPointer	
-- 
Michael A. Jackson                 400 S. Pioneer Blvd
Owner, President                   Springboro, Ohio 45066
BlueQuartz Software, LLC           EMail: mike.jackson@bluequartz.net
Voice: 937-806-1165                Web: http://www.bluequartz.net
Fax: 937-746-0783


Francois Budin wrote:
> @imikejackson <https://github.com/imikejackson> I pushed the modifications.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BlueQuartzSoftware/ITKImageProcessing/pull/4#issuecomment-224916222>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AE8TvPBx8OARlnHiYXVifp0v2qPwp-HLks5qKCYNgaJpZM4IWoge>.
>

imikejackson commented:
void itk::ImportImageContainer< 
TElementIdentifier,TElement>::SetImportPointer(...)

Can our developers just use this to tell an ITK Filter to use a 
preallocated block of memory for its output? It has been a really long 
time since I messed about with ITK so I could be just misinterpreting 

## this method?

Michael A. Jackson
BlueQuartz Software, LLC

Francois Budin wrote:

> @imikejackson https://github.com/imikejackson I pushed the modifications.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/BlueQuartzSoftware/ITKImageProcessing/pull/4#issuecomment-224916222,
> or mute the thread
> https://github.com/notifications/unsubscribe/AE8TvPBx8OARlnHiYXVifp0v2qPwp-HLks5qKCYNgaJpZM4IWoge.

imikejackson commented:
Francois,
    I have pushed a pair of filters that use part of the ITK-SIMPL 
bridge that you have created. If you want to discuss the filters that 
would be fine as I have a few questions about some of the codes.

## 

Mike Jackson  [mike.jackson@bluequartz.net]

Francois Budin wrote:

> @imikejackson https://github.com/imikejackson I pushed the modifications.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/BlueQuartzSoftware/ITKImageProcessing/pull/4#issuecomment-224916222,
> or mute the thread
> https://github.com/notifications/unsubscribe/AE8TvPBx8OARlnHiYXVifp0v2qPwp-HLks5qKCYNgaJpZM4IWoge.

fbudin69500 commented:
Thanks for pushing this filters. We can explore the limitation of the current implementation from there. I like the idea of being able to preallocate the memory for filters, and we are making sure that this is possible. Some filters have an "in place" option, which would be perfect for DREAM3D. For other filters such as the image reader, we have to make sure that allocating the memory is possible. 'ImportImageContainer' is similar to what is done in 'InPlaceDream3DDataToImageFilter', but this only works for new data that is imported in ITK, not for filters that allocate memory for the output. However, 'GraftOutput' might do the trick.
